src/xor32.rs

```diff
@@ -0,0 +1,143 @@
+///
+/// ```
+/// # extern crate alloc;
+/// use xorf::{Filter, Xor32};
+/// # use alloc::vec::Vec;
+/// # use rand::Rng;
+///
+/// # let mut rng = rand::thread_rng();
+/// const SAMPLE_SIZE: usize = 1_000_000;
+/// let keys: Vec<u64> = (0..SAMPLE_SIZE).map(|_| rng.gen()).collect();
+/// let filter = Xor32::from(&keys);
+///
+/// // no false negatives
+/// for key in keys {
+///     assert!(filter.contains(&key));
+/// }
+///
+/// // bits per entry
+/// let bpe = (filter.len() as f64) * 32.0 / (SAMPLE_SIZE as f64);
+/// assert!(bpe < 40., "Bits per entry is {}", bpe);
+///
+/// // false positive rate
+/// let false_positives: usize = (0..SAMPLE_SIZE)
+///     .map(|_| rng.gen())
+///     .filter(|n| filter.contains(n))
+///     .count();
+/// let fp_rate: f64 = (false_positives * 100) as f64 / SAMPLE_SIZE as f64;
+/// assert!(fp_rate < 0.001, "False positive rate is {}", fp_rate);
+/// ```
+///
+/// Serializing and deserializing `Xor32` filters can be enabled with the [`serde`] feature.
+///
+/// [`serde`]: http://serde.rs
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Xor32 {
+    seed: u64,
+    block_length: usize,
+    fingerprints: Box<[u32]>,
+}
+
+impl Filter<u64> for Xor32 {
+    /// Returns `true` if the filter contains the specified key. Has a false positive rate of <0.02%.
+    fn contains(&self, key: &u64) -> bool {
+        xor_contains_impl!(*key, self, fingerprint u32)
+    }
+
+    fn len(&self) -> usize {
+        self.fingerprints.len()
+    }
+
+    #[cfg(feature = "analysis")]
+    type N = u32;
+
+    #[cfg(feature = "analysis")]
+    fn fingerprints(&self) -> &[Self::N] {
+        &self.fingerprints
+    }
+}
+
+impl From<&[u64]> for Xor32 {
+    fn from(keys: &[u64]) -> Self {
+        xor_from_impl!(keys fingerprint u32)
+    }
+}
+
+impl From<&Vec<u64>> for Xor32 {
+    fn from(v: &Vec<u64>) -> Self {
+        Self::from(v.as_slice())
+    }
+}
+
+impl From<Vec<u64>> for Xor32 {
+    fn from(v: Vec<u64>) -> Self {
+        Self::from(v.as_slice())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{Filter, Xor32};
+
+    use alloc::vec::Vec;
+    use rand::Rng;
+
+    #[test]
+    fn test_initialization() {
+        const SAMPLE_SIZE: usize = 1_000_000;
+        let mut rng = rand::thread_rng();
+        let keys: Vec<u64> = (0..SAMPLE_SIZE).map(|_| rng.gen()).collect();
+
+        let filter = Xor32::from(&keys);
+
+        for key in keys {
+            assert!(filter.contains(&key));
+        }
+    }
+
+    #[test]
+    fn test_bits_per_entry() {
+        const SAMPLE_SIZE: usize = 1_000_000;
+        let mut rng = rand::thread_rng();
+        let keys: Vec<u64> = (0..SAMPLE_SIZE).map(|_| rng.gen()).collect();
+
+        let filter = Xor32::from(&keys);
+        let bpe = (filter.len() as f64) * 32.0 / (SAMPLE_SIZE as f64);
+
+        assert!(bpe < 40., "Bits per entry is {}", bpe);
+    }
+
+    #[test]
+    #[ignore]
```

Comment:
is this ignored because of the large memory cost?

Reply:
Indeed, but also the long time it takes to run as the crypto-rng is single-threaded.